test/periodic.jl

```diff
@@ -0,0 +1,369 @@
+                )
+                expected = f(2π / 23 * h)
+                @test result ≈ expected atol=1e-14
+            end
+
+            # A fall daylight-saving time change where the day has 25-hours
+            for h in 0:25
+                result = _periodic(
+                    f,
+                    ZonedDateTime(2015, 11, 1, tz"America/Winnipeg") + Hour(h),
+                    Day(1),
+                )
+                expected = f(2π / 25 * h)
+                @test result ≈ expected atol=1e-14
+            end
+        end
+
+        @testset "phase shift" begin
+            @test _periodic(sin, DateTime(1), Week(1)) == 0
+            @test _periodic(sin, DateTime(1), Week(1), Day(7)) == 0
+            @test _periodic(sin, DateTime(1), Week(1), Day(-7)) == 0
+
+            @test _periodic(sin, DateTime(2), Week(1)) ≈ 0.7818314824680298 atol=1e-14
+            @test _periodic(sin, DateTime(2), Week(1), Day(7)) ≈ 0.7818314824680298 atol=1e-14
+            @test _periodic(sin, DateTime(2), Week(1), Day(5)) ≈ 0.43388373911755823 atol=1e-14
+
+            # Using the phase shift we can make a Thursday the beginning of the week
+            # (i.e. such that sin(t) == 0). The UNIX epoch 1970-01-01 fell on Thursday.
+            # NOTE: these calculations can change depending on the period type given.
+            @test _periodic(sin, DateTime(1970), Week(1), Day(3)) == 0
+            @test _periodic(sin, DateTime(1970), Second(Week(1)), Day(5)) == 0
+
+            # Phase shifts larger than the period should not change the result
+            @test ==(
+                _periodic(sin, DateTime(1970), Day(1), Day(0)),
+                _periodic(sin, DateTime(1970), Day(1), Day(5)),
+            )
+        end
+    end
+
+    @testset "Real input" begin
+        @testset "Constructors" begin
+            @testset "Default" begin
+                p = Periodic(sin, 5, 2)
+                @test p.period == 5
+                @test p.phase_shift == 2
+            end
+
+            @testset "No phase_shift" begin
+                p = Periodic(sin, 5)
+                @test p.period == 5
+                @test p.phase_shift == 0
+            end
+        end
+
+        expected_dict = Dict(
+            sin => [
+                -0.5877852522924732,
+                -0.9510565162951535,
+                0.0,
+                0.9510565162951535,
+                0.5877852522924732,
+                -0.58778525229247
+            ],
+            cos => [
+                -0.8090169943749473,
+                0.30901699437494745,
+                1.0,
+                0.30901699437494745,
+               -0.8090169943749473,
+               -0.8090169943749475
+            ]
+        )
+
+        @testset "$f" for f in (sin, cos)
+            p = Periodic(f, 5, 2)
+
+            @test p isa Transform
+
+            @testset "Vector" begin
+                x = collect(0.:5.)
+                expected = expected_dict[f]
```

Comment:
you can pull this out of the inner loops and specify it once for all of them

Reply:
True. I think I'll have to change the name of `expected` when it is reassigned? E.g.

`expected = reshape(expected, (3, 2))`

Reply:
or just call it `_expected`